sources/packages/api/src/esdc-integration/fed-restriction-integration/fed-restriction-processing.service.ts

```diff
@@ -0,0 +1,253 @@
+import { InjectLogger } from "../../common";
+import { LoggerService } from "../../logger/logger.service";
+import { Inject, Injectable } from "@nestjs/common";
+import {
+  ConfigService,
+  FederalRestrictionService,
+  RestrictionService,
+} from "../../services";
+import { FedRestrictionIntegrationService } from "./fed-restriction-integration.service";
+import { ESDCIntegrationConfig } from "../../types";
+import * as os from "os";
+import { Connection, InsertResult } from "typeorm";
+import { FederalRestriction, Restriction } from "../../database/entities";
+import { FEDERAL_RESTRICTIONS_BULK_INSERT_AMOUNT } from "../../utilities";
+import { FedRestrictionFileRecord } from "./fed-restriction-files/fed-restriction-file-record";
+import { ProcessSFTPResponseResult } from "./fed-restriction-integration.models";
+
+/**
+ * Manages the process to import the entire snapshot of federal
+ * restrictions, that is received daily, and update the students
+ * restrictions, adding and deactivating accordingly.
+ * * This process does not affect provincial restrictions.
+ */
+@Injectable()
+export class FedRestrictionProcessingService {
+  private readonly esdcConfig: ESDCIntegrationConfig;
+  constructor(
+    @Inject("Connection") private readonly connection: Connection,
+    config: ConfigService,
+    private readonly restrictionService: RestrictionService,
+    private readonly federalRestrictionService: FederalRestrictionService,
+    private readonly integrationService: FedRestrictionIntegrationService,
+  ) {
+    this.esdcConfig = config.getConfig().ESDCIntegration;
+  }
+
+  /**
+   * Import all the federal restrictions from the SFTP and process as below:
+   * 1 - If the restriction is present on federal data and not on the
+   * student data, create a new active restriction;
+   * 2 - If the restriction in present and active on the student data
+   * but it is not present on federal data, deactivate it;
+   * 3 - If the restriction is present on federal data and it is also
+   * present and active on student data, update the updated_at only.
+   */
+  async process(): Promise<ProcessSFTPResponseResult> {
+    // Get the list of all files from SFTP ordered by file name.
+    const fileSearch = new RegExp(
+      `^${this.esdcConfig.environmentCode}CSLS.PBC.RESTR.LIST.D[\w]*\.[0-9]*`,
+      "i",
+    );
+    const filePaths = await this.integrationService.getResponseFilesFullPath(
+      this.esdcConfig.ftpResponseFolder,
+      fileSearch,
+    );
+
+    let result: ProcessSFTPResponseResult;
+    if (filePaths.length > 0) {
+      // Process only the most updated file.
+      result = await this.processAllRestrictions(
+        filePaths[filePaths.length - 1],
+      );
+      // If there are more than one file, delete it.
+      // Only the most updated file matters because it represents the entire data snapshot.
```

Comment:
So is it possible that we can get more than one file in a 24-hour interval?

Reply:
Ideally not, but if there are more than one we just need to process the most updated one since a single file represents the entire snapshot.